azure-iot-device/azure/iot/device/common/http_transport.py

```diff
@@ -82,46 +107,83 @@ def request(self, method, path, callback, body="", headers={}, query_params=""):
+        session.mount("https://", self._http_adapter)
+
+        # Format request URL
+        # TODO: URL formation should be moved to pipeline_stages_iothub_http, I believe, as
+        # depending on the operation this could have a different hostname, due to different
+        # destinations. For now this isn't a problem yet, because no possible client can
+        # support more than one HTTP operation
+        # (Device can do File Upload but NOT Method Invoke, Module can do Method Invoke and NOT file upload)
+        url = "https://{hostname}/{path}{query_params}".format(
+            hostname=self._hostname,
+            path=path,
+            query_params="?" + query_params if query_params else "",
+        )
+
         try:
-            logger.debug("creating an https connection")
-            connection = http_client.HTTPSConnection(self._hostname, context=self._ssl_context)
-            logger.debug("connecting to host tcp socket")
-            connection.connect()
-            logger.debug("connection succeeded")
-            # TODO: URL formation should be moved to pipeline_stages_iothub_http, I believe, as
-            # depending on the operation this could have a different hostname, due to different
-            # destinations. For now this isn't a problem yet, because no possible client can
-            # support more than one HTTP operation
-            # (Device can do File Upload but NOT Method Invoke, Module can do Method Inovke and NOT file upload)
-            url = "https://{hostname}/{path}{query_params}".format(
-                hostname=self._hostname,
-                path=path,
-                query_params="?" + query_params if query_params else "",
-            )
-            logger.debug("Sending Request to HTTP URL: {}".format(url))
-            logger.debug("HTTP Headers: {}".format(headers))
-            logger.debug("HTTP Body: {}".format(body))
-            connection.request(method, url, body=body, headers=headers)
-            response = connection.getresponse()
-            status_code = response.status
-            reason = response.reason
-            response_string = response.read()
-
-            logger.debug("response received")
-            logger.debug("closing connection to https host")
-            connection.close()
-            logger.debug("connection closed")
-            logger.info(
-                "https {} request sent to {}, and {} response received.".format(
-                    method, path, status_code
-                )
-            )
-            response_obj = {"status_code": status_code, "reason": reason, "resp": response_string}
-            callback(response=response_obj)
+            # Note that various configuration options are not set here due to them being set
+            # via the HTTPAdapter that was mounted at session level.
+            if method == "GET":
+                response = session.get(url, data=body, headers=headers, proxies=self._proxies)
```

Comment:
I wonder if these invocations should be perhaps running on the `http_thread` rather than the whole method. That would make the HTTPTransport work more analogously to MQTT (where Paho has it's own thread, but the MQTTTransport runs on the pipeline thread still).

This difference is why there's currently unittests failing due to timing issues FWIW